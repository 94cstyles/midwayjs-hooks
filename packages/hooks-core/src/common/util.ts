import parseArgs from 'fn-args'

export function isHooksMiddleware(fn: (...args: any[]) => any) {
  return parseArgs(fn).length === 1
}

export function extractMetadata(target: any) {
  const metadata: any = {}
  const metaKeys = Reflect.getMetadataKeys(target)
  for (const key of metaKeys) {
    metadata[key] = Reflect.getMetadata(key, target)
  }
  return metadata
}
