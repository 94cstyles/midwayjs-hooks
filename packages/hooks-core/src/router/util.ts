import { ServerRouter } from './'
import { extname, relative } from 'upath'
import _ from 'lodash'

export function getFunctionId(config: {
  router: ServerRouter
  sourceFilePath: string
  functionName: string
  isExportDefault: boolean
}) {
  const { router, sourceFilePath, functionName, isExportDefault } = config

  const rule = router.getRouteConfigBySourceFilePath(sourceFilePath)
  const lambdaDirectory = router.getApiDirectory(rule.baseDir)

  const length = router.config.routes.length
  // 多个 source 的情况下，根据各自的 lambdaDirectory 来增加前缀命名
  const relativeDirectory = length > 1 ? router.source : lambdaDirectory
  const relativePath = relative(relativeDirectory, sourceFilePath)
  // a/b/c -> a-b-c
  const id = _.kebabCase(removeExtension(relativePath))
  const name = [id, isExportDefault ? '' : `-${functionName}`].join('')
  return name.toLowerCase()
}

export function removeExtension(file: string) {
  return file.replace(extname(file), '')
}
