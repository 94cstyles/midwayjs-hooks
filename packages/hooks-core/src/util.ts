import parseArgs from 'fn-args'
import last from 'lodash/last'
import { relative, resolve, sep } from 'path'

export function isDevelopment() {
  if (
    process.env.MIDWAY_TS_MODE === 'true' ||
    /* istanbul ignore next */
    process.env.NODE_ENV === 'test' ||
    /* istanbul ignore next */
    process.env.NODE_ENV === 'development'
  ) {
    return true
  }

  /* istanbul ignore next */
  return false
}

export function useHooksMiddleware(fn: (...args: any[]) => any) {
  return (...args: any[]) => {
    /**
     * @description Hooks middleware
     * @example const middleware = (next) => { const ctx = useContext() }
     */
    if (isHooksMiddleware(fn)) {
      const next = last(args)
      return fn(next)
    }
    return fn(...args)
  }
}

export function isHooksMiddleware(fn: (...args: any[]) => any) {
  return parseArgs(fn).length === 1
}

export function isPathInside(child: string, parent: string) {
  const relation = relative(parent, child)
  return Boolean(
    relation &&
      relation !== '..' &&
      !relation.startsWith(`..${sep}`) &&
      relation !== resolve(child)
  )
}

export function extractMetadata(target: any) {
  const metadata: any = {}
  const metaKeys = Reflect.getMetadataKeys(target)
  for (const key of metaKeys) {
    metadata[key] = Reflect.getMetadata(key, target)
  }
  return metadata
}
