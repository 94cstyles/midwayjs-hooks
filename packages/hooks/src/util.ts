import createDebug from 'debug'

export const debug = createDebug('hooks')
