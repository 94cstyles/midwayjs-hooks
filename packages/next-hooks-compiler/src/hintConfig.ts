import { sync } from 'globby'
import { resolve } from 'path'

const transformers = sync(resolve(__dirname, 'plugin')).map((plugin) => ({
  name: require.resolve(plugin),
}))

export const hintConfig = {
  features: {
    tsc: {
      transformers,
    },
  },
}
