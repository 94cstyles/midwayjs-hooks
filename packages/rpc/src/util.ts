import type { RequestArgs, RequestRoute } from '@midwayjs/hooks-core'

export function parseRequestArgs<T, I = void>(requestArgs: RequestArgs<T, I>) {
  const route = requestArgs[requestArgs.length - 1] as RequestRoute<T>
  if (route.useInputMetadata) {
    return {
      route,
      inputMetadata: requestArgs[requestArgs.length - 2] as I,
      args: requestArgs.slice(0, requestArgs.length - 2),
    }
  }

  return {
    route,
    inputMetadata: null,
    args: requestArgs.slice(0, requestArgs.length - 1),
  }
}

export function buildRPCArgs(args: any[]) {
  if (!Array.isArray(args)) {
    throw new Error('args must be an array')
  }

  return {
    args,
  }
}
