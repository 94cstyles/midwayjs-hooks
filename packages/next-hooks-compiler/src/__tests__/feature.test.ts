import path from 'upath'
import { hintConfig } from '../hintConfig'
import { compileHooks } from './util'
import { clearRoutes, getFunctionsMeta } from '../routes'
import { wrap } from 'jest-snapshot-serializer-raw'

describe('features', () => {
  beforeAll(async () => {
    const fixture = path.resolve(__dirname, './fixtures/remove-underscore')
    await compileHooks(fixture, hintConfig)
  })

  it('remove underscore', () => {
    expect(wrap(JSON.stringify(getFunctionsMeta(), null, 2))).toMatchSnapshot()
  })

  afterAll(() => {
    clearRoutes()
  })
})
