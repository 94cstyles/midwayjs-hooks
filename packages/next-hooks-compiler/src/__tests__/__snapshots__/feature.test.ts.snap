// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`features remove underscore 1`] = `
{
  "arrow-hook-ref-arrow-hook-usequery": {
    "deployName": "arrow-hook-ref-arrow-hook-usequery",
    "isFunctional": true,
    "exportFunction": "useQuery",
    "sourceFile": "src/lambda/arrow-hook-ref-arrow-hook.ts",
    "sourceFilePath": "lambda/arrow-hook-ref-arrow-hook.js",
    "handler": "arrow-hook-ref-arrow-hook-usequery.handler",
    "gatewayConfig": {
      "url": "/api/arrow-hook-ref-arrow-hook/useQuery",
      "method": "POST",
      "meta": {
        "functionName": "arrow-hook-ref-arrow-hook-usequery"
      }
    },
    "event": {
      "http": {
        "basePath": "/api"
      }
    }
  },
  "arrow-hook-ref-arrow-hook-useinner": {
    "deployName": "arrow-hook-ref-arrow-hook-useinner",
    "isFunctional": true,
    "exportFunction": "useInner",
    "sourceFile": "src/lambda/arrow-hook-ref-arrow-hook.ts",
    "sourceFilePath": "lambda/arrow-hook-ref-arrow-hook.js",
    "handler": "arrow-hook-ref-arrow-hook-useinner.handler",
    "gatewayConfig": {
      "url": "/api/arrow-hook-ref-arrow-hook/useInner",
      "method": "GET",
      "meta": {
        "functionName": "arrow-hook-ref-arrow-hook-useinner"
      }
    },
    "event": {
      "http": {
        "basePath": "/api"
      }
    }
  },
  "export-arrow-hello": {
    "deployName": "export-arrow-hello",
    "isFunctional": true,
    "exportFunction": "hello",
    "sourceFile": "src/lambda/export-arrow.ts",
    "sourceFilePath": "lambda/export-arrow.js",
    "handler": "export-arrow-hello.handler",
    "gatewayConfig": {
      "url": "/api/export-arrow/hello",
      "method": "POST",
      "meta": {
        "functionName": "export-arrow-hello"
      }
    },
    "event": {
      "http": {
        "basePath": "/api"
      }
    }
  },
  "export-default-anonymous": {
    "deployName": "export-default-anonymous",
    "isFunctional": true,
    "exportFunction": "",
    "sourceFile": "src/lambda/export-default-anonymous.ts",
    "sourceFilePath": "lambda/export-default-anonymous.js",
    "handler": "export-default-anonymous.handler",
    "gatewayConfig": {
      "url": "/api/export-default-anonymous",
      "method": "GET",
      "meta": {
        "functionName": "export-default-anonymous"
      }
    },
    "event": {
      "http": {
        "basePath": "/api"
      }
    }
  },
  "export-default-func": {
    "deployName": "export-default-func",
    "isFunctional": true,
    "exportFunction": "",
    "sourceFile": "src/lambda/export-default-func.ts",
    "sourceFilePath": "lambda/export-default-func.js",
    "handler": "export-default-func.handler",
    "gatewayConfig": {
      "url": "/api/export-default-func",
      "method": "POST",
      "meta": {
        "functionName": "export-default-func"
      }
    },
    "event": {
      "http": {
        "basePath": "/api"
      }
    }
  },
  "export-default-named": {
    "deployName": "export-default-named",
    "isFunctional": true,
    "exportFunction": "",
    "sourceFile": "src/lambda/export-default-named.ts",
    "sourceFilePath": "lambda/export-default-named.js",
    "handler": "export-default-named.handler",
    "gatewayConfig": {
      "url": "/api/export-default-named",
      "method": "POST",
      "meta": {
        "functionName": "export-default-named"
      }
    },
    "event": {
      "http": {
        "basePath": "/api"
      }
    }
  }
}
`;
